x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/FieldPermissionsCache.java

```diff
@@ -67,10 +73,37 @@ FieldPermissions getFieldPermissions(String[] granted, String[] denied) {
      */
     public FieldPermissions getFieldPermissions(FieldPermissionsDefinition fieldPermissionsDefinition) {
         try {
-            return cache.computeIfAbsent(
-                fieldPermissionsDefinition,
-                (key) -> new FieldPermissions(key, FieldPermissions.initializePermittedFieldsAutomaton(key))
-            );
+            return cache.computeIfAbsent(fieldPermissionsDefinition, FieldPermissions::new);
+        } catch (ExecutionException e) {
+            if (e.getCause() instanceof ElasticsearchException es) {
+                throw es;
+            } else {
+                throw new ElasticsearchSecurityException("unable to compute field permissions", e);
+            }
+        }
+    }
+
+    public void validateFieldPermissionsWithCache(String roleName, FieldPermissionsDefinition fieldPermissionsDefinition) {
+        try {
+            // we're "abusing" the cache here a bit: to avoid excessive WARN logging, only log if the entry is not cached, i.e., on first
```

Comment:
++ FWIW With this approach we may not report all roles that have the same field permissions, but rather a first one that gets cached. Nevertheless, I think this is acceptable given that it can flood the logs.